channeldb/index_store.go

```diff
@@ -0,0 +1,264 @@
+package channeldb
+
+import (
+	"encoding/binary"
+
+	"io"
+
+	"bytes"
+
+	"github.com/boltdb/bolt"
+	"github.com/go-errors/errors"
+)
+
+var (
+	// baseIndexBucketKey base byte string name of the index store.
+	baseIndexBucketKey = []byte("indexstore")
+
+	// ErrObjectsNotFound is returned if object haven't been found by
+	// given index.
+	ErrObjectsNotFound = errors.New("objects haven't been found")
+)
+
+// Storable represents the object which might be transformed in/from
+// database specific representation. By using this interface we separate the
+// specifics of the object representation from the bolddb logic.
+type Storable interface {
+	// Decode reads the data from the byte stream and initialize the
+	// object with data.
+	Decode(r io.Reader) error
+
+	// Encode writes the internal representation of object in byte stream.
+	Encode(w io.Writer) error
+}
+
+// IndexConfig defines the configuration for the index store. ALL elements
+// within the configuration MUST be non-nil for the service to carry out its
+// duties.
+type IndexConfig struct {
+	// GetInstance returns the specific object which might be stored.
+	// Index store was designed in mind that it will be used with different
+	// types of objects.
+	GetInstance func() Storable
+}
+
+// IndexStore is the bold db storage where we assign the index to every object
+// we store in increasing manner, by doing this we preserve the order of the
+// objects and also give the ability to the user to quickly retrieve the object
+// by using this index. By using index store subsystem might initialize the map
+// of the object_id<->index during initialization and than use it for checking
+// the existence of object without touching the db itself.
+type IndexStore struct {
```

Comment:
Sure, I remove it, but the idea was to reuse this structure in other tasks. This approach allow us to decouple the business logic and (de)serialization of objects, it might allow us to reduce the level of mess in the future.